[2.5] NEXUS-5612 File listing should fail in case of IOEx

cstamas opened the pull request:
Fix for
https://issues.sonatype.org/browse/NEXUS-5612

Currently DefaultFSPeer was hiding the null returned
from File.listFiles. According to javadocs, it returns null
in two cases, a) when the File instance does not denote
a directory, or b) in case of IO error.

As case a) is checked few lines above (if check failed,
we would not get here), the null return value should
be treated as IO Error.

By throwing a generic LocalStorageExcetion, we
make this case to be treated as error.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSPeer.java:
```diff
@@ -348,6 +348,9 @@ public boolean accept( File pathname )
                                   RepositoryStringUtils.getHumanizedNameString( repository ),
                                   target.getAbsolutePath() );
+
+                throw new LocalStorageException( "Cannot list directory in repository " + repository + ", path "
```
This will result in double-logging of the same "Cannot list directory in repository" message, one from explicit log statement above, another from exception handler.

ifedorenko commented:
Why pull request commend says "related to NEXUS-5612" and commit comment does not mention NEXUS-5612 at all? Does the commit solve the problem described in NEXUS-5612?

cstamas commented:
Commit is a week old, before the issue was created. We had a phone talk with Rich where he described the problem to me, but the issue was still not created. I just wanted to save this change, resulting from that phone call.

Will remove the log and redo the commit with proper message.

cstamas committed:
NEXUS-5612: File listing should fail in case of IO problem

Currently DefaultFSPeer was hiding the null returned
from File.listFiles. According to javadocs, it returns null
in two cases, a) when the File instance does not denote
a directory, or b) in case of IO error.

As case a) is checked few lines above (if check failed,
we would not get here), the null return value should
be treated as IO Error.

By throwing a generic LocalStorageExcetion, we
make this case to be treated as error.

cstamas committed:
Merge remote-tracking branch 'origin/master' into file-list-io-error

cstamas commented:
Commit redone with Igor's proposed changes and proper message.

Code reading shows that FSPeer (and it's implementation DefaultFSPeer) is used only in FS LS (and corresponding tests) as expected.

Furthermore, code also shows that the newly thrown `LocalStorageException` will propagate to caller, and in case of Walker, will be caught and will stop the context (walk will stop).

Still, some tests proving the last (IOEx during listing + walker) are needed....

cstamas committed:
Merge remote-tracking branch 'origin/master' into file-list-io-error

cstamas committed:
Merge remote-tracking branch 'origin/master' into file-list-io-error

cstamas committed:
NEXUS-5612: Testing interaction with WalkerProcessor and FSPeer

Especially the case when FSPeer#listItems method throws
LocalStorageException.

The Walk must be stopped, after this, no processor
should have onCollectionExit invoked.

This UT actually simulates the case when due to
too many open files, FSPeer made Walker think
that collection is empty, while it was not.

DefaultFSPeer is already fixed in prev commit,
this UT only ensures that Walker, Context and
Processors behave correctly in this case.

cstamas committed:
NEXUS-5612: Remove UT since tag

Remnant from copied skeleton UT.

adreghiciu commented:
+1

nabcos commented:
+1

adreghiciu commented:
Started a CI build: https://bamboo.zion.sonatype.com/browse/NXO-OSSF20-1

adreghiciu commented:
CI :+1:
